### PR TITLE
Authentication API Integration

### PR DESCRIPTION
**Pull request by eriyezab**
1. Create endpoint to check log in status of users.
2. Add front end functionality to hide comment form from logged out users and provide either a log in or logout link to the respective users.

### CODE REVIEW

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 1]
```diff
@@ -27,18 +27,42 @@
...
+
+  public String getEmail() {
+    return this.email;
+  }
+
+  public String getMessage() {
+    return this.email;
+  }
+
+  public long timestamp() {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: Any reason for not using the same convention and calling this getTimestamp?

File: portfolio/src/main/java/com/google/sps/data/UserStatus.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+/**
+ * Class representing the logged in status of the user.
+ *
+ * <p>Note: The private variables in this class are converted into JSON.
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/data/UserStatus.java]: I didn't quite understand this comment. Is the conversion important to mention for some reason? If so, could you please expand the comment to include that.

File: portfolio/src/main/java/com/google/sps/data/UserStatus.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ */
+public class UserStatus {
+
+  /** The logged in status of the user */
+  private boolean loggedIn;
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/data/UserStatus.java]: You can call this isLoggedIn to be more clear.

File: portfolio/src/main/java/com/google/sps/data/UserStatus.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+  /** The url that the user clicks to log in or log out depending on status. */
+  private String url;
+
+  public UserStatus(boolean loggedIn, String url) {
+    this.loggedIn = loggedIn;
+    this.url = url;
+  }
+  
+  public boolean getLoggedIn() {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/data/UserStatus.java]: You can name this isLoggedIn as well. Not sure if the java style guide mentions this. But apparently the general JavaBean specification does: https://stackoverflow.com/questions/1118261/what-is-the-correct-syntax-for-is-variable-getter-setters-in-a-pojo-class

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 5]
```diff
@@ -68,8 +72,17 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Check if the user is logged in and if not redirect them. If so get the email
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: Comments need to be punctuated properly. Please add '.' at the end.

File: portfolio/src/main/webapp/comments.js [hunk 6]
```diff
@@ -1,5 +1,8 @@
 const FILTERS = document.getElementById("filters");
 const COMMENTS_LIST_DOC_ELEMENT = document.getElementById('comments-list');
+const USER_DIV = document.getElementById("users");
+const LOG_IN = "Click here to log in: ";
```
Comment (shrutip90) [inline portfolio/src/main/webapp/comments.js]: The constant name doesn't look complete. Maybe use LOG_IN_PROMPT or LOG_IN_MESSAGE.

File: portfolio/src/main/webapp/comments.js [hunk 7]
```diff
@@ -1,5 +1,8 @@
...
+const LOG_OUT = "Click here to log out: ";
```
Comment (shrutip90) [inline portfolio/src/main/webapp/comments.js]: ditto

File: portfolio/src/main/webapp/comments.js [hunk 8]
```diff
@@ -48,8 +51,15 @@ function getComments() {
     console.log(comments);
     for(i = 0; i < comments.length; ++i) {
       let listNode = document.createElement("LI");
+      let displayName;
+      if(comments[i].name) {
+        displayName = comments[i].name;
+      } else {
+        displayName = comments[i].email;
+      }
```
Comment (shrutip90) [inline portfolio/src/main/webapp/comments.js]: Can you replace this with an assignment using a ternary operator?

File: portfolio/src/main/webapp/comments.js [hunk 9]
```diff
@@ -78,4 +88,19 @@ function deleteAllComments() {
   }
 }
 
+function getUserStatus() {
+  fetch("/user")
+  .then(res => res.json())
+  .then((userStatus) => {
+    if(userStatus.loggedIn){
```
Comment (shrutip90) [inline portfolio/src/main/webapp/comments.js]: How is this code able to access the private variable?

**Review by shrutip90 — CHANGES_REQUESTED** [review]

File: portfolio/src/main/webapp/comments.js (hunk 9 above)
Comment (eriyezab) [reply portfolio/src/main/webapp/comments.js]: The UserStatus class is sent from the server as json so I am accessing the loggedIn(now isLoggedIn) member from the json it received. I updated the UserStatus class javadoc/comment to explain it more but let me know if it needs more explanation.

File: portfolio/src/main/java/com/google/sps/data/Comment.java [hunk 10]
```diff
@@ -27,18 +27,42 @@
...
+
+  public String getMessage() {
+    return this.email;
+  }
+
+  public long getTimestamp() {
+    return this.timestamp;
+  }
+
+
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/data/Comment.java]: You can remove these if they are not intentional - double blank lines at the end of file.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 11]
```diff
@@ -51,40 +54,50 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+    // Send json to server.
     response.setContentType("applications/json;");
     response.getWriter().println(json);
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Check if the user is logged in and if not redirect them. If so get the email.
+    String email;
+    if(USER.isUserLoggedIn()) {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: if statement should be followed by a space. See https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace.

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 12]
```diff
@@ -51,40 +54,50 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
+
     // Get input from the form.
-    String name = getParameter(request, "name", "Anonymous");
+    String name = getParameter(request, "name", null); 
     String message = getParameter(request, "message", null);
     long timestamp = System.currentTimeMillis();
 
-    // If the message was empty then do not add to datastore 
+    // If the message was empty then do not add to datastore.
     if(message == null) {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: ditto

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 13]
```diff
@@ -106,6 +119,8 @@ private Query makeQueryFromParams(HttpServletRequest request) {
     String sortBy;
     if(sortValue.equals("name")) {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: ditto

File: portfolio/src/main/java/com/google/sps/servlets/DataServlet.java [hunk 14]
```diff
@@ -106,6 +119,8 @@ private Query makeQueryFromParams(HttpServletRequest request) {
...
       sortBy = "name";
+    } else if(sortValue.equals("email")) {
```
Comment (shrutip90) [inline portfolio/src/main/java/com/google/sps/servlets/DataServlet.java]: ditto

File: portfolio/src/main/webapp/comments.js [hunk 15]
```diff
@@ -78,4 +82,19 @@ function deleteAllComments() {
   }
 }
 
+function getUserStatus() {
+  fetch("/user")
+  .then(res => res.json())
+  .then((userStatus) => {
+    if(userStatus.isLoggedIn){
```
Comment (shrutip90) [inline portfolio/src/main/webapp/comments.js]: ditto

**Review by shrutip90 — APPROVED** [review]

**Pull request merged**